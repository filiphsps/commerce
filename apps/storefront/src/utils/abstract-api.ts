/* eslint-disable unused-imports/no-unused-vars */
import type { Nullable, OnlineShop } from '@nordcom/commerce-db';

import { DocumentTransform } from '@apollo/client';
import { visit } from 'graphql';

import type { Locale } from '@/utils/locale';
import type { ApolloClient, TypedDocumentNode } from '@apollo/client';

export type OmitTypeName<T> = Omit<T, '__typename'>;

export type ApiOptions = { api: AbstractApi };

export type AbstractApi<Q = any> = {
    locale: () => Locale;
    shop: () => OnlineShop;
    query: <T>(
        query: Q,
        variables?: Record<string, string | number | boolean | object | Array<string | number | object> | null>,
        options?: {
            fetchPolicy?: RequestCache;
            tags?: string[];
            revalidate?: number;
        }
    ) => Promise<{ data: T | null; errors: readonly any[] | undefined }>;
};
export type AbstractApiBuilder<K, Q> = ({
    api,
    locale,
    shop
}: {
    api: K;
    locale: Locale;
    shop: OnlineShop;
}) => AbstractApi<Q>;

export type AbstractShopifyApolloApiBuilder<Q> = AbstractApiBuilder<ApolloClient<any>, Q>;

export function buildCacheTagArray(shop: OnlineShop, locale: Locale, tags: string[]) {
    // TODO: change `shopify` tag based on the api we're using.
    return ['shopify', `shopify.${shop.id}`, shop.domain, locale.code, ...tags];
}

/**
 * Creates an AbstractApiBuilder for Shopify Apollo APIs.
 *
 * @todo TODO: Improve the type safety of all `AbstractApi` implementations.
 *
 * @param {object} options - The api options.
 * @param {ApolloClient<any>} options.api - The Apollo client to use.
 * @param {Locale} options.locale - The locale to use.
 * @param {OnlineShop} options.shop - The locale to use.
 * @returns {AbstractApiBuilder} The AbstractApiBuilder.
 */
export const ApiBuilder: AbstractShopifyApolloApiBuilder<TypedDocumentNode<any, any>> = ({ api, locale, shop }) => ({
    locale: () => locale,
    shop: () => shop,
    query: async (query, variables = {}, { tags = [], revalidate = undefined, fetchPolicy = undefined } = {}) => {
        const { data, errors, error } = await api.query({
            query,
            //fetchPolicy,
            context: {
                fetchOptions: {
                    cache: fetchPolicy ?? 'no-cache',
                    next: {
                        revalidate: revalidate ?? undefined,
                        tags: buildCacheTagArray(shop, locale, tags)
                    }
                }
            },
            variables: {
                language: locale.language,
                country: locale.country,
                ...variables
            }
        });

        return { data: data || null, errors, error };
    }
});

/**
 * @todo TODO: This should be replaced with a generalized shopify parser.
 *       Preferably one that we can use to output in whatever format we want.
 */
export const cleanShopifyHtml = (html: string | unknown): Nullable<string> => {
    if (typeof html !== 'string' || !html) return null;
    let out = (html as string) || '';

    // Remove all non-breaking spaces and replace them with normal spaces.
    // TODO: This is a hacky solution. We should write a proper shopify parser.
    out = out.replaceAll(/ /g, ' ').replaceAll('\u00A0', ' ');

    // Replace some of the more common unicode characters with their HTML.
    out = out
        .replaceAll('”', '&rdquo;')
        .replaceAll('“', '&ldquo;')
        .replaceAll('‘', '&lsquo;')
        .replaceAll('’', '&rsquo;')
        .replaceAll('…', '&hellip;');

    // Trim the preceding and trailing whitespace.
    out = out.trim();

    return out;
};

/**
 * A GraphQL document transform that adds the `@inContext` directive to all queries.
 *
 * THIS IS (or at least seems to be) A REALLY HACKY SOLUTION. USE WITH CAUTION. :^)
 *
 * @todo TODO: This should be placed in a separate file or even a separate package.
 * @todo TODO: Need to add tests for this.
 * @todo TODO: This currently adds duplicate variables if they already exist.
 */
export const shopifyContextTransform = new DocumentTransform((document) => {
    // Add Shopify's `@inContext` directive to all queries.
    const transformedDocument = visit(document, {
        OperationDefinition(node) {
            return {
                ...node,
                variableDefinitions: [
                    ...(node.variableDefinitions || []),
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'country'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'CountryCode'
                            }
                        }
                    },
                    {
                        kind: 'VariableDefinition',
                        variable: {
                            kind: 'Variable',
                            name: {
                                kind: 'Name',
                                value: 'language'
                            }
                        },
                        type: {
                            kind: 'NamedType',
                            name: {
                                kind: 'Name',
                                value: 'LanguageCode'
                            }
                        }
                    }
                ],
                directives: [
                    ...(node.directives || []),
                    {
                        kind: 'Directive',
                        name: {
                            kind: 'Name',
                            value: 'inContext'
                        },
                        arguments: [
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'country'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'country'
                                    }
                                }
                            },
                            {
                                kind: 'Argument',
                                name: {
                                    kind: 'Name',
                                    value: 'language'
                                },
                                value: {
                                    kind: 'Variable',
                                    name: {
                                        kind: 'Name',
                                        value: 'language'
                                    }
                                }
                            }
                        ]
                    }
                ]
            };
        }
    });
    return transformedDocument;
});

export type ApiReturn<T> = [T, undefined] | [undefined, Error];
