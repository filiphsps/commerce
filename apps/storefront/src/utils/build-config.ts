/* c8 ignore start */
export const BuildConfig = {
    environment: process.env.NODE_ENV,

    shopify: {
        api: '2024-07' as const
    }
};
/* c8 ignore stop */
