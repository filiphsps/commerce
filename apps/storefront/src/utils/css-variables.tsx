import { Shop } from '@nordcom/commerce-db';
import { TodoError } from '@nordcom/commerce-errors';

import { colord, extend } from 'colord';
import a11yPlugin from 'colord/plugins/a11y';

extend([a11yPlugin]);

// TODO: Generalize this
export const getBrandingColors = async (domain: string) => {
    const shop = await Shop.findByDomain(domain, { sensitiveData: true });
    if (shop.design.accents.length <= 0) {
        throw new TodoError();
    }
    const accents = shop.design.accents;

    // TODO: Deal with variants.
    const primary = accents
        .filter(({ type }) => type === 'primary')
        .sort((a, b) => (colord(a.color).luminance() < colord(b.color).luminance() ? -1 : 1))[0];
    const secondary = accents
        .filter(({ type }) => type === 'secondary')
        .sort((a, b) => (colord(a.color).luminance() < colord(b.color).luminance() ? -1 : 1))[0];

    return {
        primary,
        secondary
    };
};

const CssVariablesProvider = async ({ domain }: { domain: string }) => {
    const branding = await getBrandingColors(domain);

    // TODO: Background and foreground colors.
    return (
        <style suppressHydrationWarning={true}>{`
        :root {
            --color-background: #fefefe;
            --color-foreground: #101418;

            --color-accent-primary: ${branding.primary.color};
            --color-accent-primary-text: ${branding.primary.foreground};
            --color-accent-primary-light: ${colord(branding.primary.color).lighten(0.115).saturate(0.15).toHex()};
            --color-accent-primary-dark: ${colord(branding.primary.color).darken(0.05).toHex()};

            --color-accent-secondary: ${branding.secondary.color};
            --color-accent-secondary-text: ${branding.secondary.foreground};
            --color-accent-secondary-light: ${colord(branding.secondary.color).lighten(0.195).saturate(0.15).toHex()};
            --color-accent-secondary-dark: ${colord(branding.secondary.color).darken(0.15).toHex()};

            /* TODO: Remove these legacy variables. */
            --accent-primary: var(--color-accent-primary);
            --accent-primary-light: var(--color-accent-primary-light);
            --accent-primary-dark: var(--color-accent-primary-dark);
            --accent-secondary: var(--color-accent-secondary);
            --accent-secondary-light: var(--color-accent-secondary-light);
            --accent-secondary-dark: var(--color-accent-secondary-dark);
        }
    `}</style>
    );
};

CssVariablesProvider.displayName = 'Nordcom.CssVariablesProvider';
export { CssVariablesProvider };
