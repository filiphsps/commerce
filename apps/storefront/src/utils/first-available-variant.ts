import { NotFoundError } from '@nordcom/commerce-errors';

import type { Product, ProductVariant } from '@/api/product';

/**
 * Find the first available and most suitable variant for a specific product.
 *
 * @param {Product} product - The product data.
 * @returns {ProductVariant} The recommended variant based on availability and price.
 */
export const firstAvailableVariant = (product?: Product | null): ProductVariant | undefined => {
    // 1. Make sure we got a product passed to us.
    if (!product?.variants) {
        return undefined;
    }

    const variants: ProductVariant[] =
        ((product.variants.edges as any)
            ? product.variants.edges.map(({ node: variant }) => variant)
            : (product.variants as any)) || [];

    if (variants.length <= 0) {
        throw new NotFoundError(`"product.variant"`);
    }

    // 2. Check if the last variant is available.
    if (variants.at(-1)?.availableForSale) {
        // 2.1. If it is, return it.
        return variants.at(-1)!;
    }
    // 2.2. If it isn't, continue.

    // 3. Loop through the variants in reverse to get the more expensive variants first.
    for (let i = variants.length - 1; i >= 0; i--) {
        const variant = variants[i] as (typeof variants)[number] | undefined;

        // 3.1. If the variant is not available, continue.
        if (!variant?.availableForSale) {
            continue;
        }

        // 3.2. If the variant is available, return it.
        return variant;
    }

    // 4. If no variants are available, return the most expensive one.
    return variants.at(-1)!;
};
