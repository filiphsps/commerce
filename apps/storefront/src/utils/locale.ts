import React from 'react';

import { TodoError, UnknownLocaleError } from '@nordcom/commerce-errors';

import ConvertUnits from 'convert-units';

import type english from '@/i18n/en.json';
import type { StoreModel } from '@/models/StoreModel';
import type { CountryCode, CurrencyCode, LanguageCode, WeightUnit } from '@shopify/hydrogen-react/storefront-api-types';
import type { ReactNode } from 'react';

export type { CountryCode, CurrencyCode, LanguageCode };
export type Code = `${Lowercase<LanguageCode>}-${CountryCode}` | Lowercase<LanguageCode>;

type SerializableLocale = {
    /**
     * @deprecated Use `code` instead.
     */
    locale: Code;

    code: Code;
    language: LanguageCode;
    country?: CountryCode;
};

/**
 * A locale.
 */
export class Locale implements SerializableLocale {
    public code!: Code;
    public language: LanguageCode;
    public country?: CountryCode;

    /**
     * @deprecated Use `code` instead.
     */
    public locale!: Code;

    private constructor({ language, country }: { language: LanguageCode; country?: CountryCode }) {
        this.language = language;
        this.country = country;

        if (typeof country !== 'undefined') {
            this.code = `${language.toLowerCase()}-${country.toUpperCase()}` as Code;
        } else {
            this.code = `${language.toLowerCase()}` as Code;
        }

        // TODO: Remove `locale` when `locale` is removed from `SerializableLocale`.
        this.locale = this.code;
    }

    /**
     * The default locale.
     *
     * @todo TODO: This should be tenant configurable.
     */
    static get default(): Readonly<SerializableLocale> {
        // FIXME: Don't hardcode `en-US` as the fallback.
        return Locale.from('en-US' as Code);
    }

    /**
     * Get the current locale.
     */
    static get current(): Readonly<SerializableLocale> {
        if (typeof window === 'undefined') {
            // TODO: This should be based on the current shop.
            return Locale.default;
        }

        if (window.locale) {
            return Locale.from(window.locale);
        }

        // Get the locale code from the first path segment.
        const code = window.location.pathname.split('/')[1];
        if (!code) {
            throw new UnknownLocaleError(`Invalid locale: "${code}"`);
        }

        return Locale.from(code);
    }

    /**
     * Check if a locale is configured as the default locale.
     *
     * @param {Locale} locale - The locale
     * @returns {boolean} `true` if the locale is the default locale, otherwise `false`.
     */
    static isDefault(locale: Locale): boolean {
        return locale ? locale.code === Locale.default.code : false;
    }

    /**
     * Convert data to a `Locale`.
     *
     * @param {Code | any} data - The basis to create the locale from.
     * @returns {Locale} The immutable locale.
     */
    static from(data: { language: LanguageCode; country?: CountryCode } | Code | string) {
        // We can only pass pure objects to the client.
        const wrap = (locale: Locale) =>
            Object.freeze(Object.fromEntries(Object.entries(locale)) as SerializableLocale);

        if (typeof data === 'string') {
            const code = data.toUpperCase() as Uppercase<Code>;

            if (!code || code.length < 2 || code.length > 5 || (code.length !== 2 && !code.includes('-'))) {
                throw new UnknownLocaleError(`Invalid locale: "${data}"`);
            }

            if (code.length === 2) {
                return wrap(new Locale({ language: code as LanguageCode }));
            }

            const [language, country] = code.split('-') as [LanguageCode, CountryCode?];
            return wrap(new Locale({ language, country }));
        } else {
            if (!data) throw new UnknownLocaleError(`Invalid locale: "${data}"`);

            const { language, country } = data;
            if (language.length !== 2 || (typeof country !== 'undefined' && language.length !== 2)) {
                throw new UnknownLocaleError(`Invalid locale: "${data}"`);
            }

            return wrap(new Locale({ language, country }));
        }
    }
}

/**
 * Converts a locale string to a `CountryCode`.
 *
 * @deprecated Use {@link Locale.from} instead.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {CountryCode} `CountryCode` string.
 */
export const NextLocaleToCountry = (locale?: string): CountryCode => {
    return (locale ? Locale.from(locale) : Locale.default).country || ('US' as const);
};

/**
 * Converts a locale string to a `LanguageCode`.
 *
 * @deprecated Use {@link Locale.from} instead.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {LanguageCode} `LanguageCode` string.
 */
export const NextLocaleToLanguage = (locale?: string): LanguageCode => {
    return (locale ? Locale.from(locale) : Locale.default).language;
};

/**
 * Converts a locale string to a `CurrencyCode`.
 *
 * @param {string} locale - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {CurrencyCode} `CurrencyCode` string.
 */
export const NextLocaleToCurrency = ({ country, store }: { country: CountryCode; store: StoreModel }): CurrencyCode =>
    (store.payment?.countries?.find(({ isoCode }) => isoCode === country)?.currency.isoCode! || 'USD') as CurrencyCode;

/**
 * Converts a locale string to a Locale.
 *
 * @note If the locale is invalid, the default locale will be used.
 *       the default locale is defined in `Config.i18n.default` and
 *       is not tenant configurable at the moment.
 *
 * @deprecated Use {@link Locale.from} instead.
 *
 * @param {string} code - The `ISO 639-1` + `ISO 3166-1 Alpha-2` or pure `ISO 639-1` locale string.
 * @returns {Locale} `Locale` object.
 */
export const NextLocaleToLocale = (code: string): Locale | null => {
    // Legacy handling.
    if (!code || code.length < 2 || code.length > 5 || (code.length > 2 && !code.includes('-'))) {
        return null;
    }

    // Legacy handling.
    if (code.length === 2) {
        throw new TodoError();
    }

    return Locale.from(code);
};

export type DeepKeys<T> = T extends object
    ? {
          [K in keyof T]-?: K extends string | number ? `${T[K] extends object ? DeepKeys<T[K]> : K}` : never;
      }[keyof T]
    : never;

/**
 * A dictionary, uses `en.json` as the base for type-safety.
 */
export type LocaleDictionary = typeof english;
export type LocaleDictionaryScope = Lowercase<keyof LocaleDictionary>;
export type LocaleDictionaryKey = Lowercase<DeepKeys<LocaleDictionary>>;

type TranslationLiteral = string | number | boolean | ReactNode;
/**
 * Returns a translation function for a given scope and dictionary.
 *
 * > NOTE: This is a very simple implementation of a translation function.
 *
 * @param {string} scope - The scope of the translation.
 * @param {LocaleDictionary} dictionary - The dictionary to use for the translation.
 * @returns {({ t: (key: string, ...literals: TranslationLiteral[]) => string })} The translation function.
 */
export const useTranslation = (scope: LocaleDictionaryScope, dictionary?: LocaleDictionary) => {
    return {
        // FIXME: Fix return type.
        t: <T extends LocaleDictionaryKey, L extends TranslationLiteral[]>(key: T, ...literals: L): string => {
            const string: string = (dictionary as any)?.[scope]?.[key] || key;

            if (!literals || literals.length === 0) {
                return string as string;
            }

            const placeholderRegex = /\{([^}]+)\}/g;
            const parts: (string | ReactNode)[] = [];

            let match;
            let lastIndex = 0;
            while ((match = placeholderRegex.exec(string)) !== null) {
                parts.push(string.substring(lastIndex, match.index));
                const index = parseInt(match[1], 10);
                parts.push(literals[index]);
                lastIndex = match.index + match[0].length;
            }
            parts.push(string.substring(lastIndex));

            const partsWithKeys = parts.map((part, index) =>
                React.isValidElement(part) ? { ...{ key: index }, ...part } : part
            );
            return partsWithKeys.some((part) => React.isValidElement(part)) ? partsWithKeys : (parts.join('') as any);
        }
    };
};

export const ConvertToLocalMeasurementSystem = ({
    locale,
    weight,
    weightUnit
}: {
    locale: Locale;
    weight: number;
    weightUnit: WeightUnit;
}): string => {
    const weightUnitToConvertUnits = (unit: WeightUnit) => {
        switch (unit.toLowerCase()) {
            case 'grams':
                return 'g';
            case 'kilograms':
                return 'kg';
            case 'ounces':
                return 'oz';
            case 'pounds':
                return 'lb';

            // TODO: Handle this; which should never possibly actually occur.
            default: {
                console.warn(`Unknown weight unit: ${unit}, defaulting to grams.`);
                return 'g';
            }
        }
    };
    // TODO: Support more than just US here, because apparently there's a lot
    //        more countries out there using imperial.
    const metric = locale.country ? locale.country.toLowerCase() !== 'us' : true;
    const unit = weightUnitToConvertUnits(weightUnit);

    // TODO: Do this properly.
    const targetUnit = metric ? 'g' : 'oz';

    if (unit !== targetUnit) weight = ConvertUnits(weight).from(unit).to(targetUnit);

    let res = ((Math.round(weight) * 100) / 100).toFixed(metric ? 0 : 2).toString();
    if (res.endsWith('.00')) res = res.slice(0, -3);

    return `${res}${targetUnit}`;
};
