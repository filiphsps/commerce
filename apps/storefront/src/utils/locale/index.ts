export * from './locale';
export * from './weight';
