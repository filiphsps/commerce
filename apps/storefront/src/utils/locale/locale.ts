import React from 'react';

import { UnknownLocaleError } from '@nordcom/commerce-errors';

import type english from '@/i18n/en.json';
import type { CountryCode, CurrencyCode, LanguageCode } from '@shopify/hydrogen-react/storefront-api-types';
import type { ReactNode } from 'react';

export type { CountryCode, CurrencyCode, LanguageCode };
export type Code = `${Lowercase<LanguageCode>}-${CountryCode}` | Lowercase<LanguageCode>;

type SerializableLocale = {
    /**
     * @deprecated Use `code` instead.
     */
    locale: Code;

    code: Code;
    language: LanguageCode;
    country?: CountryCode;
};

/**
 * A locale.
 */
export class Locale implements SerializableLocale {
    public code!: Code;
    public language: LanguageCode;
    public country?: CountryCode;

    /**
     * @deprecated Use `code` instead.
     */
    public locale!: Code;

    private constructor({ language, country }: { language: LanguageCode; country?: CountryCode }) {
        this.language = language;
        this.country = country;

        if (typeof country !== 'undefined') {
            this.code = `${language.toLowerCase()}-${country.toUpperCase()}` as Code;
        } else {
            this.code = `${language.toLowerCase()}` as Code;
        }

        // TODO: Remove `locale` when `locale` is removed from `SerializableLocale`.
        this.locale = this.code;
    }

    /**
     * The default locale.
     */
    static get default(): Readonly<SerializableLocale> {
        // FIXME: Don't hardcode `en-US` as the fallback.
        return Locale.from('en-US' as Code);
    }

    /**
     * Get the current locale.
     */
    static get current(): Readonly<SerializableLocale> {
        if (typeof window === 'undefined') {
            // TODO: This should be based on the current shop.
            return Locale.default;
        }

        if (window.locale) {
            return Locale.from(window.locale);
        }

        // Get the locale code from the first path segment.
        const code = window.location.pathname.split('/')[1];
        if (!code) {
            throw new UnknownLocaleError(`Invalid locale: "${code}"`);
        }

        return Locale.from(code);
    }

    /**
     * Check if a locale is configured as the default locale.
     *
     * @param {Locale} locale - The locale
     * @returns {boolean} `true` if the locale is the default locale, otherwise `false`.
     */
    static isDefault(locale: Locale = Locale.default): boolean {
        return locale.code === Locale.default.code;
    }

    /**
     * Convert data to a `Locale`.
     *
     * @param {Code | any} data - The basis to create the locale from.
     * @returns {Locale} The immutable locale.
     */
    static from(data: { language: LanguageCode; country?: CountryCode } | Code | string) {
        // We can only pass pure objects to the client.
        const wrap = (locale: Locale) =>
            Object.freeze(Object.fromEntries(Object.entries(locale)) as SerializableLocale);

        if (typeof data === 'string') {
            const code = data.toUpperCase() as Uppercase<Code>;

            if (!(code as any) || code.length < 2 || code.length > 5 || (code.length !== 2 && !code.includes('-'))) {
                throw new UnknownLocaleError(`Invalid locale: "${data}"`);
            }

            if (code.length === 2) {
                return wrap(new Locale({ language: code as LanguageCode }));
            }

            const [language, country] = code.split('-') as [LanguageCode, CountryCode?];
            return wrap(new Locale({ language, country }));
        } else {
            if (!(data as any)) {
                throw new UnknownLocaleError(`Invalid locale: "${data}"`);
            }

            const { language, country } = data;
            if ((language as any).length !== 2 || (!!country && (language as any).length !== 2)) {
                throw new UnknownLocaleError(`Invalid locale: "${data}"`);
            }

            return wrap(new Locale({ language, country }));
        }
    }
}

export type DeepKeys<T> = T extends object
    ? {
          [K in keyof T]-?: K extends string | number ? `${T[K] extends object ? DeepKeys<T[K]> : K}` : never;
      }[keyof T]
    : never;

/**
 * A dictionary, uses `en.json` as the base for type-safety.
 */
export type LocaleDictionary = typeof english;
export type LocaleDictionaryScope = Lowercase<keyof LocaleDictionary>;
export type LocaleDictionaryKey = Lowercase<DeepKeys<LocaleDictionary>>;

type TranslationLiteral = string | number | boolean | ReactNode;
/**
 * Returns a translation function for a given scope and dictionary.
 *
 * > NOTE: This is a very simple implementation of a translation function.
 *
 * @param {string} scope - The scope of the translation.
 * @param {LocaleDictionary} dictionary - The dictionary to use for the translation.
 * @returns {({ t: (key: string, ...literals: TranslationLiteral[]) => string })} The translation function.
 */
export const getTranslations = (scope: LocaleDictionaryScope, dictionary?: LocaleDictionary) => {
    return {
        // FIXME: Fix return type.
        t: <T extends LocaleDictionaryKey, L extends TranslationLiteral[]>(key: T, ...literals: L): string => {
            const string: string = (dictionary as any)?.[scope]?.[key] || key;

            if (((literals as any)?.length || 0) <= 0) {
                return string as string;
            }

            const placeholderRegex = /\{([^}]+)\}/g;
            const parts: (string | ReactNode)[] = [];

            let match;
            let lastIndex = 0;
            while ((match = placeholderRegex.exec(string)) !== null) {
                parts.push(string.substring(lastIndex, match.index));
                const index = parseInt(match[1], 10);
                parts.push(literals[index]);
                lastIndex = match.index + match[0].length;
            }
            parts.push(string.substring(lastIndex));

            const partsWithKeys = parts.map((part, index) =>
                React.isValidElement(part) ? { ...{ key: index }, ...part } : part
            );
            return partsWithKeys.some((part) => React.isValidElement(part)) ? partsWithKeys : (parts.join('') as any);
        }
    };
};

/**
 * Checks if the option name is a size option.
 * @param {string} name - The name of the option.
 * @returns {boolean} `true` if the option is a size option, otherwise `false`.
 */
export const isSizeOption = (name: string): boolean =>
    [
        'size', // English.
        'tamaño', // Spanish.
        'größe', // German.
        'storlek' // Swedish, Norwegian, Danish.
    ].includes(name.toLowerCase());

type CapitalizeOptions = {
    everyWord?: boolean;
    lowerCase?: boolean;
};
/**
 *
 * @param {string} string - The string to capitalize.
 * @param {CapitalizeOptions} options - The options.
 * @param {boolean} [options.everyWord=false] - Whether to capitalize every word in the string.
 * @param {boolean} [options.lowerCase=true] - Whether to convert the rest of the string to lowercase.
 * @returns {string} The capitalized string.
 */
export function capitalize(string: string, { everyWord = false, lowerCase = true }: CapitalizeOptions = {}): string {
    if (everyWord) {
        return string
            .split(' ')
            .map((word) => capitalize(word, { lowerCase }))
            .join(' ');
    }

    return string.charAt(0).toUpperCase() + (lowerCase ? string.slice(1).toLowerCase() : string.slice(1));
}
