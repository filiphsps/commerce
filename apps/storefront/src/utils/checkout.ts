import type { OnlineShop } from '@nordcom/commerce-db';
import { InvalidCartError, UnknownCommerceProviderError } from '@nordcom/commerce-errors';

import { productToMerchantsCenterId } from '@/utils/merchants-center-id';
import { safeParseFloat } from '@/utils/pricing';

import type { Locale } from '@/utils/locale';
import type { TrackableContextValue } from '@/utils/trackable';
import type { CartWithActions } from '@shopify/hydrogen-react';
import type { CartLine } from '@shopify/hydrogen-react/storefront-api-types';

// Const hacky workaround for ga4 cross-domain
// taken from StackOverflow
export const getCrossDomainLinkerParameter = (domain: string) => {
    // create form element, give it an action, make it hidden and prevent the submit event
    const formNode = document.createElement('form') as any;
    formNode.action = `https://checkout.${domain}`; // TODO: This should be dependant on the tenant.
    formNode.style.opacity = '0';
    formNode.addEventListener('submit', (event: any) => {
        event.preventDefault();
    });

    // create a button node, make it type=submit and append it to the form
    const buttonNode = document.createElement('button') as any;
    buttonNode.type = 'submit';
    formNode.append(buttonNode);

    // append the form (and button) to the DOM
    document.body.append(formNode);

    // trigger a click on the button node to submit the form
    buttonNode.click();

    // check for the input[name=_gl] hidden input in the form (if decoration worked)
    const _glNode = formNode.querySelector('input[name="_gl"]') as any;

    if (_glNode) {
        return _glNode.value as string;
    }

    console.warn(`Could not find _gl input in checkout form with action "${formNode.action}"`);
    return null;
};

export const Checkout = async ({
    shop,
    locale,
    cart,
    trackable
}: {
    shop: OnlineShop;
    locale: Locale;
    cart: CartWithActions;
    trackable: TrackableContextValue;
}) => {
    if (shop.commerceProvider.type !== 'shopify') {
        throw new UnknownCommerceProviderError();
    }

    if (typeof (cart as any) === 'undefined' || !(cart as any)) {
        throw new InvalidCartError('Cart is undefined or null');
    } else if (!cart.totalQuantity || cart.totalQuantity <= 0 || !cart.lines) {
        throw new InvalidCartError('Cart is empty');
    } else if (!cart.checkoutUrl) {
        throw new InvalidCartError('Cart is missing checkoutUrl');
    }

    const url = cart.checkoutUrl.replace(/[A-Za-z0-9\-\_]+.\.myshopify\.com/, shop.commerceProvider.domain);

    try {
        trackable.postEvent('begin_checkout', {
            gtm: {
                ecommerce: {
                    currency: cart.cost?.totalAmount?.currencyCode!,
                    value: safeParseFloat(undefined, cart.cost?.totalAmount?.amount),
                    items: (cart.lines.filter((_) => _) as CartLine[]).map((line) => ({
                        item_id: productToMerchantsCenterId({
                            locale: locale,
                            product: {
                                productGid: line.merchandise!.product!.id,
                                variantGid: line.merchandise!.id
                            } as any
                        }),
                        item_name: line.merchandise.product.title,
                        item_variant: line.merchandise.title,
                        item_brand: line.merchandise.product.vendor,
                        item_category: line.merchandise.product.productType || undefined,
                        sku: line.merchandise.sku || undefined,
                        product_id: line.merchandise!.product!.id,
                        variant_id: line.merchandise!.id,
                        currency: line.merchandise.price.currencyCode!,
                        price: safeParseFloat(undefined, line.merchandise.price.amount),
                        quantity: line.quantity
                    }))
                }
            }
        });
    } catch (error: unknown) {
        console.error(error);
    }

    const ga4 = getCrossDomainLinkerParameter(shop.domain);
    window.location.href = `${url}${ga4 ? `${url.includes('?') ? '&' : '?'}_gl=${ga4}` : ''}`;
};
