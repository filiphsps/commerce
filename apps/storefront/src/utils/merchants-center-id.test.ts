import { describe, expect, it } from 'vitest';

import { Locale } from '@/utils/locale';
import { productToMerchantsCenterId } from '@/utils/merchants-center-id';

describe('utils', () => {
    describe('MerchantsCenterId', () => {
        it(`should generate a Merchants Center product ID based on the product's Shopify GIDs`, () => {
            const product = {
                productGid: 'gid://shopify/Product/123',
                variantGid: 'gid://shopify/ProductVariant/456'
            } as any;

            const locale = Locale.from('en-US')!;

            const merchantsCenterId = productToMerchantsCenterId({
                locale,
                product
            });

            expect(merchantsCenterId).toBe('shopify_US_123_456');
            //expect(merchantsCenterId).toMatchSnapshot();
        });

        it(`should generate a Merchants Center product ID based on the product's Shopify GIDs`, () => {
            const product = {
                productGid: 'gid://shopify/Product/123',
                variantGid: 'gid://shopify/ProductVariant/456'
            } as any;

            const locale = Locale.from('fr-CA')!;

            const merchantsCenterId = productToMerchantsCenterId({
                locale,
                product
            });

            expect(merchantsCenterId).toBe('shopify_CA_123_456');
            //expect(merchantsCenterId).toMatchSnapshot();
        });
    });
});
