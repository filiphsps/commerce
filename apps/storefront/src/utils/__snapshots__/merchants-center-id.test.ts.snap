// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > MerchantsCenterId > should generate a Merchants Center product ID based on the product's Shopify GIDs 1`] = `"shopify_US_123_456"`;

exports[`utils > MerchantsCenterId > should generate a Merchants Center product ID based on the product's Shopify GIDs 2`] = `"shopify_CA_123_456"`;
